library: add modules

guits opened the pull request:
This adds various modules:

- cephadm_bootstrap: for bootstrapping new clusters with cephadm,
- ceph_config: for getting/setting ceph option,
- ceph_orch_host: for adding/removing hosts,
- ceph_orch_apply: for applying service spec,
- ceph_orch_daemon: for starting/stopping services,
- ceph_orch: intended to be used as an ansible wrapper in order to run any `ceph orch` commands via ansible playbooks.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
library: add ceph_orch module

This adds the `ceph_orch` module which is intended
to be used as an ansible wrapper in order to run any
`ceph orch|cephadm` commands via ansible playbooks.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
tests: add mypy env in tox configuration

in order to run mypy tests from tox.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits commented:
jenkins test deploy_and_purge

guits commented:
jenkins test deploy_and_purge

guits commented:
jenkins test deploy_and_purge

guits commented:
jenkins test deploy_and_purge

guits committed:
tests: update tox.ini

add explicitly envlist in tox section

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
tests: use cephadm_bootstrap module

this makes the ci use the cephadm_bootstrap module in order
to test it.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
tests: use a specific sha1

shaman API seems to have issue at the moment, let's use a specific
sha1 in order to unlock the CI.

to be reverted asap.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
tests: update ansible.cfg

This updates the ansible.cfg for the CI.
It adds the paths to library and module_utils.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
tests: use ceph_orch module

this makes the ci use the ceph_orch module in order
to test it.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits commented on library/cephadm_bootstrap.py:
```diff
@@ -0,0 +1,260 @@
+
+    if registry_url and registry_username and registry_password:
+        cmd.extend(['--registry-url', registry_url,
+                    '--registry-username', registry_username,
+                    '--registry-password', registry_password])
+
+    if registry_json:
+        cmd.extend(['--registry-json', registry_json])
+
+    if ssh_user:
+        cmd.extend(['--ssh-user', ssh_user])
+
+    if ssh_config:
+        cmd.extend(['--ssh-config', ssh_config])
+
+    if module.check_mode:
+        exit_module(
+            module=module,
+            out='',
+            rc=0,
+            cmd=cmd,
+            err='',
+            startd=startd,
+            changed=False
+        )
+    else:
+        rc, out, err = module.run_command(cmd)
+        exit_module(
+            module=module,
+            out=out,
+            rc=rc,
+            cmd=cmd,
+            err=err,
+            startd=startd,
+            changed=True
+        )
+
+
+if __name__ == '__main__':
+    main()
```
with the very recent changes in `ceph_orch.py` and `ceph_common.py` we can probably get rid of this module now.

guits committed:
tests: remove legacy task

this task isn't needed, let's remove it from that playbook.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
library: add ceph_orch documentation

This adds the documentation in the module ceph_orch.py

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits commented:
jenkins test deploy_and_purge

pcuzner reviewed (commented):
I'm not sure about the goals for this module as it stands? 

For example, what does this do over and above a simple command or shell invocation to one of the admin nodes? With this module you're still expecting the ansible user to pass the correct syntax of the required cephadm command.

The command passed isn't validated, so can fail on the target instead of failing syntax as the playbook is validated.

As it stands this requires the ansible user to be a cephadm CLI expert too, right.

For this to appeal to ansible users should it be structured/delivered as a collection in galaxy, and the code in this repo just reference the collection?

guits committed:
library: add cephadm_bootstrap module

add cephadm_bootstrap module.

(took from ceph-ansible project, dim ^^)

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
bootstrap: make the module idempotent

the module fails when it's run against a cluster already deployed.
this commits makes it skip and no throw an error if a cluster is
already running.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
library: add ceph_orch_host module

This adds the ceph_orch_host module.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
tests: use ceph_orch_host module

Use ceph_orch_host module in order to test it in the CI.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
library: add ceph_orch_apply module

This adds the ceph_orch_apply module

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
tests: use ceph_orch_apply module

This makes the CI use the ceph_orch_apply module.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
library: move `build_base_cmd` to module_utils

This moves the `build_base_cmd()` to ceph_common in module_utils.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
library: add ceph_orch_daemon module

this adds the `ceph_orch_daemon` module.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
tests: add flake8 check

This adds flake8 check.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
tests: rename deploy_and_purge env

rename an env (`deploy_and_purge` to `tox`)

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
library: add 'fsid' param

This adds the `fsid` parameter to `ceph_orch_*` modules.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
library: add ceph_config module

This adds the ceph_config module.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
library: add/fix check_mode support

This adds/fixes the check_mode support in all modules.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

ktdreyer commented:
Will the RPM package still include all the new files you're adding here? I think we need to add this new `library` directory to this list in `cephadm-ansible.spec.in`;

```
for f in ansible.cfg *.yml ceph_defaults; do
...
```

guits commented:
> Will the RPM package still include all the new files you're adding here?

yes

> I think we need to add this new `library` directory to this list in `cephadm-ansible.spec.in`;
> 
> ```
> for f in ansible.cfg *.yml ceph_defaults; do
> ...
> ```

and yes 🙂

guits committed:
tests: support multi-distro testing

testing refactor in order to support multi-distro testing

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
pkg/spec: add library and module_utils

This makes the rpm specfile install these two directories.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
doc: add rst documentation

This adds the documentation about the overall project.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits commented on doc/source/index.rst:
```diff
@@ -0,0 +1,344 @@
+.. ----------------
+
+.. .. toctree::
+..    :maxdepth: 1
```
```suggestion
```

adk3798 commented on doc/source/index.rst:
```diff
@@ -0,0 +1,418 @@
+* cephadm-clients.yml: Setting up client hosts
+* cephadm-purge-cluster.yml: Remove a Ceph cluster
+
+In addition to that, several ansible modules are provided in order to let people writing their own playbooks.
```
```suggestion
Additionally, several ansible modules are provided in order to help users write their own playbooks.
```

adk3798 commented on doc/source/index.rst:
```diff
@@ -0,0 +1,418 @@
+-----------
+
+**admin host**
+  A host where the admin keyring and ceph config file is present. Although the admin host and the bootstrap host are usually the same host, it is possible to have multiple admin hosts later.\
```
```suggestion
  A host where the admin keyring and ceph config file is present. Although the admin host and the bootstrap host are usually the same host, it is possible to have multiple admin hosts later. \
```
Not sure if it's just on my end but the spaces weren't rendering here when I tested locally
![image](https://user-images.githubusercontent.com/47704447/169865767-73dc8860-76bb-408a-a524-7f89ca2ddf93.png)

adk3798 commented on doc/source/index.rst:
```diff
@@ -0,0 +1,418 @@
+  The host where any cephadm-ansible playbook is run.
+
+**bootstrap host**
+  The host where the ceph cluster will start. Unless you pass ``--skip-admin-label`` option to ``ceph bootstram`` command, this host will get the admin keyring and the ceph config file present, therefore, it should be considered as an 'admin host'.
```
```suggestion
  The host where the ceph cluster will start. Unless you pass the ``--skip-admin-label`` option to the ``ceph bootstrap`` command, this host will receive the admin keyring and ceph config file and should be considered an 'admin host'.
```

adk3798 commented on doc/source/index.rst:
```diff
@@ -0,0 +1,418 @@
+
+**bootstrap host**
+  The host where the ceph cluster will start. Unless you pass ``--skip-admin-label`` option to ``ceph bootstram`` command, this host will get the admin keyring and the ceph config file present, therefore, it should be considered as an 'admin host'.
+  This hosts should be present in the group ``[admin]`` in the ansible inventory.
```
```suggestion
  This host should be present in the group ``[admin]`` in the ansible inventory.
```

adk3798 commented on doc/source/index.rst:
```diff
@@ -0,0 +1,418 @@
+
+Ansible inventory
+-----------------
+The ansible inventory is a file where all the hosts intended to be part of the ceph cluster will be listed.\
```
seeing same as my other comment with the space not being added after the \

adk3798 commented on doc/source/index.rst:
```diff
@@ -0,0 +1,418 @@
+The ansible inventory is a file where all the hosts intended to be part of the ceph cluster will be listed.\
+The most common format are INI or YAML.
+
+Although you probably want to keep it as simple as possible, you can organize your inventory and create groups, `cephadm-ansible` won't make any difference except for the following requirements:
```
```suggestion
Although you probably want to keep it as simple as possible, you can organize your inventory and create groups, `cephadm-ansible` won't differentiate except for the following requirements:
```

adk3798 commented on doc/source/index.rst:
```diff
@@ -0,0 +1,418 @@
+Although you probably want to keep it as simple as possible, you can organize your inventory and create groups, `cephadm-ansible` won't make any difference except for the following requirements:
+
+* Client hosts must be defined in a dedicated group ``[clients]``.
+* Both ``cephadm-purge-cluster.yml`` and ``cephadm-clients.yml`` playbooks requires a group ``[admin]`` with at least one admin host (usually it will be the bootstrap node).
```
```suggestion
* Both the ``cephadm-purge-cluster.yml`` and ``cephadm-clients.yml`` playbooks require a group ``[admin]`` with at least one admin host (usually it will be the bootstrap node).
```

adk3798 commented on doc/source/index.rst:
```diff
@@ -0,0 +1,418 @@
+
+.. note:: the name of the client group can be changed. In that case you have to set the variable `client_group`.
+
+Otherwise, you can create groups such as ``[monitors]``, ``[osds]``, ``[rgws]``, that might help you keep clarity in your inventory file and probably ease the ``--limit`` usage if you plan to use it to target a group of node only.
```
```suggestion
Otherwise, you can create groups such as ``[monitors]``, ``[osds]``, ``[rgws]``, that might help you keep clarity in your inventory file and ease the ``--limit`` usage if you plan to use it to target a specific group of nodes only.
```

adk3798 commented on doc/source/index.rst:
```diff
@@ -0,0 +1,418 @@
+=================
+
+This playbook configures the Ceph repository.
+It also installs some prerequisites (podman, lvm2, chronyd, cephadm, ...)
```
do we want to note the issues with certain distros with this? Specifically thinking of the issues we saw with Fedora a while ago.

guits replied:
yes, I think we should add a specific section for that

adk3798 commented on doc/source/index.rst:
```diff
@@ -0,0 +1,418 @@
+   node123
+
+
+Then you can run the playbook as shown above.
```
does this section about clients stuff need to be in this section? It seems like maybe it belongs in the section for the clients playbook.

guits replied:
yes, it must be a leftover from when the clients setup was part of the preflight playbook

adk3798 commented on doc/source/index.rst:
```diff
@@ -0,0 +1,418 @@
+
+keyring
+~~~~~~~~
+**description**
+  The full path name of the keyring file on the host (which should be admin[0]) which holds the key for the client to use
+
+conf
+~~~~
+**description**
+  The full path name of the conf file on the (which should be admin[0]) host to use (undefined will trigger a minimal conf)
```
are there default values for the conf and keyring here?

guits replied:
for `conf` : 

> (undefined will trigger a minimal conf)

`keyring` is a mandatory parameter, no default value

adk3798 commented on doc/source/index.rst:
```diff
@@ -0,0 +1,418 @@
+Introduction
+============
+
+cephadm-ansible provides several modules so it is possible to write playbooks in order to have some automation on top of cephadm/ceph orch.
```
```suggestion
cephadm-ansible provides several modules to make it easier to write playbooks around cephadm/ceph orch.
```

adk3798 commented on doc/source/index.rst:
```diff
@@ -0,0 +1,418 @@
+
+cephadm-ansible provides several modules so it is possible to write playbooks in order to have some automation on top of cephadm/ceph orch.
+The idea is to let you write your own playbooks so you can have a solution that really fits with your needs.
+At the moment only most important tasks are supported. It means that any operation uncovered would have to be done either with either ``command`` or ``shell`` Ansible tasks in your playbook.
```
```suggestion
At the moment only the most important tasks are supported. This means that any operation not covered would have to be done either with either the ``command`` or ``shell`` Ansible tasks in your playbook.
```

adk3798 commented on doc/source/index.rst:
```diff
@@ -0,0 +1,418 @@
+=======
+
+This shows how the supported modules can be used in a playbook.
+This doesn't cover the pre-requisites steps (preflight, ...) so it implies all requirements satisfied (podman, lvm2,...).
```
```suggestion
This doesn't cover the pre-requisites steps (preflight, ...) so it assumes all requirements are satisfied (podman, lvm2,...).
```

adk3798 commented on doc/source/index.rst:
```diff
@@ -0,0 +1,418 @@
+
+This shows how the supported modules can be used in a playbook.
+This doesn't cover the pre-requisites steps (preflight, ...) so it implies all requirements satisfied (podman, lvm2,...).
+It assumes your "bootstrap host" (or "admin host") can ssh other hosts with root user without password.
```
```suggestion
It assumes your "bootstrap host" (or "admin host") can ssh to other hosts with root user without password.
```

adk3798 commented on doc/source/index.rst:
```diff
@@ -0,0 +1,418 @@
+            data_devices:
+                  all: true
+
+   - name: change osd_memory_target option for osd.0
```
note that in some newer version we turn on osd memory autotuning by default during bootstrap so if somebody sets this config option it might just get overwritten. Not sure if that means a different example would be better here.

guits replied:
> Not sure if that means a different example would be better here.

I think so

adk3798 commented on doc/source/index.rst:
```diff
@@ -0,0 +1,418 @@
+       - name: apply osd spec
+         ceph_orch_apply:
+         spec: |
+            service_type: osd
+            service_id: osd
+            placement:
+            label: osd
+            spec:
+            data_devices:
+                  all: true
```
```suggestion
            service_type: osd
            service_id: osd
            placement:
                  label: osd
            spec:
                  data_devices:
                        all: true
```
The spec will fail to apply without the indentation I think

adk3798 reviewed (commented):
a lot of small stuff. Docs generally look good

guits committed:
tests: add unit tests

This adds some unit tests for module_utils and the
module `cephadm_bootstrap.py`.

More tests will be added for the other modules in a separate PR.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

guits committed:
tests: check pkgs are installed on client hosts

Check `ceph-common` and `chrony` are installed on client
hosts.

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>
